### DISCUSSION TITLE
Custom self-sign root CA and LDAPS AD SSL TLS

### DISCUSSION THREAD

**Discussion by BuhtigithuB** [requarks/wiki#3429] [Help / Questions]
**Question**
Describe in details what you are trying to achieve and where do you need guidance.

Having difficulty with LDAP**S**... I suspect that since AD is using locally signed local domain self-signed root CA, I have to install it into certs bundle... Although I am not sure if I can install it into alpine/ubuntu certs bundle or Node.js bundle...

Tried to do it like so : 

Dockerfile
```
FROM requarks/wiki:2

ADD custom_root_ca_sha256.crt /custom_root_ca_sha256.crt
ENV NODE_EXTRA_CA_CERTS="/custom_root_ca_sha256.crt"
```

Didn't help...

Tried with : 

```
FROM requarks/wiki:2

RUN apk --no-cache add ca-certificates
ADD custom_root_ca_sha256.crt /usr/local/share/ca-certificates/custom_root_ca_sha256.crt
RUN update-ca-certificates
```

No luck neither...

Also, it is not clear to me what is the Dockerfile of the requarks/wiki:2 docker hub container image: https://github.com/Requarks/wiki/blob/dev/dev/build/Dockerfile **???**

Like how Mattermost docker-compose does related to that matter of allowing tweaks to base image : 

https://github.com/mattermost/mattermost-docker

As you use docker-compose.yml build, so you can customize what you need easily from the Dockerfile of the base image...

**Comment by NGPixel** [comment]
The dockerfile used for the build is:
https://github.com/Requarks/wiki/blob/b15d93d993c64d8481338af89ad82086529f0b8c/dev/build/Dockerfile#L23-L54

**Comment by BuhtigithuB** [comment]
Will try with that... Will report back... Thanks

**Comment by BuhtigithuB** [comment]
Having difficulty with assets... Not cloned the repo yet, but I was searching for their location, not sure where they are coming from...

**Comment by BuhtigithuB** [comment]
@NGPixel, not sure how to proceed... Git repo not seems to have continuous integration setup that I know about... So I can't glance any insight on how to build the wiki container image... :disappointed:

**Comment by NGPixel** [comment]
https://docs.requarks.io/dev/build-process

**Comment by BuhtigithuB** [comment]
HA! Didn't dig enough in the documentation... I was scratching my head as to how such a great project wouldn't have a proper CI in place :smile: 

Thanks

**Comment by BuhtigithuB** [comment]
Been a little epic to finally figured out how to build wiki.js main container image but I guess I succeed as I get docker build to succeed and been able to run the custom built image in docker-compose stack... Next release will look nice ;)

Although I make really sure that  my custom root ca get in the custom image built and that I set NODE_EXTRA_CA_CERTS environment variable to point to the alpine certs bundle that I also check and contains a copy of custom root ca as follow : 

```
FROM node:12-alpine
LABEL maintainer="requarks.io"


RUN apk add bash curl git openssh gnupg sqlite ca-certificates --no-cache && \
    mkdir -p /wiki && \
    mkdir -p /logs && \
    mkdir -p /wiki/data/content && \
    chown -R node:node /wiki /logs

ADD leddartech_root_sha256.crt /usr/local/share/ca-certificates/leddartech_root_sha256.crt
RUN update-ca-certificates
ENV NODE_EXTRA_CA_CERTS="/etc/ssl/certs/ca-certificates.crt"

...
```


And still have this in the logs : 

```
wiki_1   | 2020-08-26T02:27:57.057Z [MASTER] error: Authentication Strategy LDAP / Active Directory (ldap): [ FAILED ]
wiki_1   | 2020-08-26T02:27:57.057Z [MASTER] error: ENOENT: no such file or directory, open
```

And this error in the UI : 

```
Invalid authentication provider.
```

LDAP:389 authenticate just fine, but I can't just get LDAPS:636 to work... To my knowledge I have AD with SSL certificate no need for TLS certificates with path to certificate to verify, so plain "Use TLS" check box ON shall suffice...

The only remaining thing I can think of is to put my custom root ca somewhere else for Node.js to pick it up...

??

Thanks

**Comment by BuhtigithuB** [comment]
Realize that NODE_EXTRA_CA_CERTS is not set properly rebuild try with : 

```
RUN echo 'export NODE_EXTRA_CA_CERTS="/etc/ssl/certs/ca-certificates.crt"' >> /home/node/.bashrc
```

Try also that : 

```
UN echo 'export NODE_EXTRA_CA_CERTS="/etc/ssl/certs/ca-certificates.crt"' | cat - /etc/profile > /etc/new_profile && mv /etc/new_profile /etc/profile
```

Just can can't get the environment variable to be setted when I check like this : 

```
docker run -ti --rm wikijs_requark_2:latest bash -c "echo $NODE_EXTRA_CA_CERTS"
```

I get the PATH printed out just fine...

I am a bit puzzled...

**Comment by NGPixel** [comment]
The dev branch won't work with any other authentication than local at the moment. Use the latest branch instead if you need LDAP working.

**Comment by BuhtigithuB** [comment]
ok!

I had tried base on that answer https://stackoverflow.com/a/53180691 also... Just can't get environment variable to get created... I am not use to Alpine... Will try latest...

Thanks for replying that late...

**Comment by BuhtigithuB** [comment]
Tried with the git latest branch custom build and with requark/wiki:latest pulled container some old!!

Will be offline for a couples of days, will be back next monday or tuesday...

Thanks for your help

**Comment by BuhtigithuB** [comment]
Some esoteric readings : 

https://stackoverflow.com/questions/29283040/how-to-add-custom-certificate-authority-ca-to-nodejs/39972054#39972054

https://stackoverflow.com/questions/44459971/nodejs-environment-variable-node-extra-ca-certs

https://stackoverflow.com/questions/14262986/node-js-hostname-ip-doesnt-match-certificates-altnames

https://stackoverflow.com/questions/31861109/tls-what-exactly-does-rejectunauthorized-mean-for-me/31862256#31862256
